### PR TITLE
implement plotly_editor_traits

### PR DESCRIPTION
**Pull request by nicolaskruchten**
proposed solution to #258 (more robust interactions with children)

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
This isn't particularly revolutionary or beautiful, but it does give us looser coupling than straight equality checks on `child.type` or indexOf checks on `child.type.displayName`...

**Comment by bpostlethwaite** [comment]
Nope not a revolution but just the :ticket: ! Neatly captures a pattern. Any tests warranted? It would be nice to have something fail if one of the traits fails to get passed through during some future refactor.

**Comment by nicolaskruchten** [comment]
This PR just leverages the existing tests basically.

File: src/components/containers/PanelHeader.js [hunk 1]
```diff
@@ -74,13 +55,18 @@ class PanelHeader extends Component {
...
+PanelHeader.contextTypes = {
+  layout: PropTypes.object,
+  onUpdate: PropTypes.func,
+  updateContainer: PropTypes.func,
+};
+
 PanelHeader.propTypes = {
-  action: PropTypes.string,
+  addAction: PropTypes.object,
   allowCollapse: PropTypes.bool,
```
Comment (VeraZab) [inline src/components/containers/PanelHeader.js]: this is really nice! but should this 'action' stay more general, maybe it's not always going to be an 'addAction'

File: src/components/containers/Panel.js [hunk 2]
```diff
@@ -100,40 +86,32 @@ class Panel extends Component {
...
-        return child;
-      });
+      );
 
       return (
         <div className={bem('panel')}>
           <PanelHeader
-            action={action}
+            addAction={addAction}
             allowCollapse={nbOfFolds > 1}
```
Comment (VeraZab) [inline src/components/containers/Panel.js]: same here

File: src/components/containers/PanelHeader.js (hunk 1 above)
Comment (nicolaskruchten) [reply src/components/containers/PanelHeader.js]: At the moment it's pretty well coupled to being `add` if we hardcode the `<PlusIcon>` like we do... Should I generalize the icon as well?

Comment (VeraZab) [reply src/components/containers/PanelHeader.js]: yeah, you're right, I didn't notice the icon, we should be generalizing that too then if going that route.

Well, ok, let's generalize when we need to generalize. I don't foresee us needing other actions in the panel header, other than add, anytime soon

**Comment by VeraZab** [comment]
ok, well, I'm 💃  on this

**Pull request merged**